include/deal.II/lac/qr.h

```diff
@@ -0,0 +1,863 @@
+ *   journal    = {ACM Trans. Math. Softw.},
+ *   year       = {1990},
+ *   volume     = {16},
+ *   number     = {4},
+ *   pages      = {369--377},
+ *   month      = dec,
+ *   issn       = {0098-3500},
+ *   acmid      = {98291},
+ *   address    = {New York, NY, USA},
+ *   doi        = {10.1145/98267.98291},
+ *   issue_date = {Dec. 1990},
+ *   numpages   = {9},
+ *   publisher  = {ACM},
+ *   url        = {http://doi.acm.org/10.1145/98267.98291},
+ *  }
+ * @endcode
+ */
+template <typename VectorType>
+class QR : public BaseQR<VectorType>
+{
+  // clang-format on
+public:
+  /**
+   * Number type for R matrix.
+   */
+  typedef typename VectorType::value_type Number;
+
+  /**
+   * Default constructor.
+   */
+  QR();
+
+  /**
+   * Destructor
+   */
+  virtual ~QR() = default;
+
+  /**
+   * @copyfrom BaseQR::append_column
+   *
+   * @note Currently this function always returns <code>true</code>.
+   */
+  virtual bool
+  append_column(const VectorType &column);
+
+  /**
+   * Remove first column and update QR factorization.
+   *
+   * Starting from the given QR decomposition
+   * $QR= A = [a_1\,\dots a_n], \quad a_i \in {\mathbb R}^m$
+   * we aim at computing factorization of
+   * $\tilde Q \tilde R= \tilde A = [a_2\,\dots a_n], \quad a_i \in {\mathbb
+   * R}^m$.
+   *
+   * The standard approach is to partition $R$ as
+   * \f[
+   * R =
+   * \begin{bmatrix}
+   * r_{11} & w^T \\
+   * 0      & R_{33}
+   * \end{bmatrix}
+   * \f]
+   * It then follows that
+   * \f[
+   * Q^T \tilde A =
+   * \begin{bmatrix}
+   * 0 & w^T \\
+   * 0 & R_{33}
+   * \end{bmatrix}
+   * \f]
+   * is upper Hessenberg where unwanted sub-diagonal elements can be
+   * zeroed by a sequence of Givens rotations.
+   *
+   * Note that $\tilde R^T \tilde R = \tilde A^T \tilde A$,
+   * where the RHS is included in $A^T A = R^T R$. Therefore
+   * $\tilde R$ can be obtained by Cholesky decomposition.
+   */
+  virtual void
+  remove_column(const unsigned int k = 0);
+
+  virtual void
+  multiply_with_Q(VectorType &y, const Vector<Number> &x) const;
+
+  virtual void
+  multiply_with_QT(Vector<Number> &y, const VectorType &x) const;
+
+  virtual void
+  multiply_with_A(VectorType &y, const Vector<Number> &x) const;
+
+  virtual void
+  multiply_with_AT(Vector<Number> &y, const VectorType &x) const;
+
+private:
+  /**
+   * Apply givens rotation in plane @p i @p k to @p Q and @p R so that
+   * <code>R(k,k)</code> is zeroed.
+   *
+   * See Chapter 5.1.9 of Golub 2013, Matrix computations.
+   */
+  void
+  apply_givens_rotation(const unsigned int i, const unsigned int k);
+
+  /**
+   * Temporary vector needed to do Givens rotation of Q.
+   */
+  VectorType tmp;
+};
+
+
+
+/**
+ * The class to obtain $R$ triangular matrix of $A=QR$ factorization
+ * together with the matrix $A$ itself. Orthonormal $Q$ is not stored
+ * explicitly, but can be obtained from ${R^T}^{-1}A^T = Q^T$, thus
```

Comment:
That seems like a complicated formula, but it's really also `Q = A R^{-1}`. Is your formula easier to implement, or why did you choose it?

Reply:
> Is your formula easier to implement, or why did you choose it?

actually I use both for `multiply_with_QT` and `multiply_with_Q`. I will update the docu.